experimental/Schemes/CoherentSheaves.jl

```diff
@@ -162,31 +163,373 @@ identifications given by the glueings in the `default_covering`.
       return MU
     end
 
+    function production_func(
+        F::AbsPreSheaf,
+        U::SimplifiedSpec
+      )
+      V = original(U)
+      MV = F(V)
+      rho = OOX(V, U)
+      MU, phi = change_base_ring(rho, MV)
+      add_incoming_restriction!(F, V, MU, phi)
+      return MU
+    end
+
     ### Production of the restriction maps; to be cached
     function restriction_func(F::AbsPreSheaf, V::AbsSpec, U::AbsSpec)
-      MV = F(V)
-      MU = F(U)
-      # There are two cases: Either U is a PrincipalOpenSubset of V, or U 
-      # is a PrincipalOpenSubset of another affine_chart W. In both cases, 
-      # we need to find W (which equals V in the first case) and use the transition 
-      # matrix for the changes between these two sets. 
-      VV = V
-      while VV isa PrincipalOpenSubset
-          VV = ambient_scheme(VV)
+      if any(W->(W === U), affine_charts(X)) && any(W->(W === V), affine_charts(X))
+        MV = F(V)
+        MU = F(U)
+        A = MG[(V, U)] # The transition matrix
+        UU, _ = glueing_domains(default_covering(X)[U, V])
+        psi = OOX(UU, U) # Needs to exist by the checks of is_open_func, even though 
+        # in general UU ⊂ U!
+        return hom(MV, MU, [sum([psi(A[i, j]) * MU[j] for j in 1:ngens(MU)]) for i in 1:ngens(MV)], rho)
```

Comment:
This line will throw an error if `ngens(MU)` is zero, e.g. for the zero module. And maybe it is not type stable.
The reason is that the empty sum is not well defined, because julia does not know which zero element to choose.
You can fix this by providing the keyword  `init=zero(MU)` .